File: framework/ruby/src/soda/leetcode/tree.rb
module Soda end

module Soda::Leetcode
  
class TreeNode
  attr_accessor :val, :left, :right
  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

class TreeFactory
  def self.create(data)
    return nil if data.empty?

    root = TreeNode.new(data[0])
    qu = [root]
    index = 1
    while index < data.size
      node = qu.shift
      if data[index]
        node.left = TreeNode.new(data[index])
        qu.push(node.left)
      end
      index += 1
      if index == data.size
        break
      end
      if data[index]
        node.right = TreeNode.new(data[index])
        qu.push(node.right)
      end
      index += 1
    end
    root
  end

  def self.dump(root)
    return [] unless root
    res = []
    qu = [root]
    until qu.empty?
      node = qu.shift
      if node
        res << node.val
        qu << node.left
        qu << node.right
      else
        res << nil
      end
    end
    until res[-1]
      res.pop
    end
    res
  end
end

end

File: framework/js/src/soda/leetcode/tree.js
function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}

class TreeFactory {
    static create(data) {
        if (data.length === 0) {
            return null;
        }
        let root = new TreeNode(data[0]);
        let qu = [root];
        for (let i = 1; i < data.length; ) {
            let node = qu.shift();
            if (data[i]) {
                node.left = new TreeNode(data[i]);
                qu.push(node.left);
            }
            if (++i === data.length) {
                break;
            }
            if (data[i]) {
                node.right = new TreeNode(data[i]);
                qu.push(node.right);
            }
            ++i;
        }
        return root;
    }

    static dump(root) {
        if (!root) {
            return [];
        }
        let res = [];
        let qu = [root];
        while (qu.length > 0) {
            let node = qu.shift();
            if (node) {
                res.push(node.val);
                qu.push(node.left);
                qu.push(node.right);
            } else {
                res.push(null);
            }
        }
        while (!res.at(-1)) {
            res.pop();
        }
        return res;
    }
}

exports.TreeNode = TreeNode;
exports.TreeFactory = TreeFactory;
